Fabric.Terminology.SqlServer/Persistence/SqlValueSetRepository.cs

```diff
@@ -143,17 +164,114 @@ public IValueSet GetValueSet(string valueSetUniqueId, IEnumerable<string> codeSy
             var systemCodes = codeSystemCodes as string[] ?? codeSystemCodes.ToArray();
             if (systemCodes.Any())
             {
-                var codevsid = this.SharedContext.ValueSetCodes
-                    .Where(code => systemCodes.Contains(code.CodeSystemCD))
-                    .Select(code => code.ValueSetID)
+                var codevsid = this.SharedContext.ValueSetCodes.Where(code => systemCodes.Contains(code.CodeSystemCD))
+                    .Select(code => code.ValueSetUniqueID)
                     .Distinct();
 
-                dtos = dtos.Join(codevsid, id => id.ValueSetID, sa => sa, (id, sa) => id);
+                dtos = dtos.Join(codevsid, id => id.ValueSetUniqueID, sa => sa, (id, sa) => id);
             }
 
             return this.CreatePagedCollectionAsync(dtos, pagerSettings, systemCodes, includeAllValueSetCodes);
         }
 
+        public Attempt<IValueSet> Add(IValueSet valueSet)
+        {
+            if (!valueSet.IsCustom)
+            {
+                return Attempt<IValueSet>.Failed(new ValueSetOperationException("Only custom Value Sets may be created or updated."));
+            }
+
+            valueSet.SetIdsForCustomInsert();
+
+            var valueSetDto = valueSet.AsDto();
+            var codeDtos = valueSet.ValueSetCodes.Select(code => code.AsDto());
+
+            this.ClientTermContext.ChangeTracker.AutoDetectChangesEnabled = false;
+            using (var transaction = this.ClientTermContext.Database.BeginTransaction())
+            {
+                try
+                {
+                    this.ClientTermContext.ValueSetDescriptions.Add(valueSetDto);
+                    this.ClientTermContext.ValueSetCodes.AddRange(codeDtos);
+                    this.ClientTermContext.SaveChanges();
```

Comment:
A buddy of mine out of the UK has done some pretty clever stuff with EF bulk inserts (millions of records).  Was planning on hitting him up for a walkthrough.